trunk/source/CACSL2BoogieTranslator/src/de/uni_freiburg/informatik/ultimate/cdt/translation/implementation/base/standardfunctions/StandardFunctionHandler.java

```diff
@@ -1305,6 +1302,86 @@ private Result handleAtomicExchange(final IDispatcher main, final IASTFunctionCa
 		return builder.build();
 	}
 
+	// https://gcc.gnu.org/onlinedocs/gcc/_005f_005fatomic-Builtins.html#index-_005f_005fatomic_005fcompare_005fexchange_005fn
+	// https://en.cppreference.com/w/c/atomic/atomic_compare_exchange
+	//
+	// The implementation below generates Boogie code that roughly follows this schema, where success and ptr_val are
+	// auxiliary variables:
+	//
+	// @formatter:off
+	// (evaluate arguments)
+	// atomic {
+	//  if weak
+	//    havoc success
+	//  else
+	//    success := true
+	//  ptr_val := read(ptr)
+	//  if success && ptr_val == read(expected):
+	//    write(read(desired), ptr)
+	//  else
+	//    success := false
+	// }
+	// if !success
+	//   write(ptr_val, expected)
+	// return success
+	// @formatter:on
+	private Result handleAtomicCompareExchange(final IDispatcher main, final IASTFunctionCallExpression node,

```

Comment:
If I understand the specification correctly, then the only difference between `__atomic_compare_exchange_n` and `__atomic_compare_exchange` is that the latter one has a pointer as `desired` that is simply dereferenced and written into `ptr`. Is it possible to refactor this method to call `handleAtomicCompareExchangeN`?

Reply:
:+1: done